### PR TITLE
Intake and Ejector Configurator

### PR DESCRIPTION
**Pull request by Dhruv-0-Arora**
### Description
Implementation of configuring the positioning of a robot's intake and ejector. Uses raycasting to detect the node (very basic implementation of detecting node) and saves the information. Allows user to scale the size of the sphere that spawns as well.

### Features
- [x] Configure pickup zone
- [x] Configure ejector
- [x] Create physics sensor for pickup zone
- [x] Attach ejectable game piece
- [x] Eject ejectable game piece

### Robot with configured pickup zone
image

### Robot holding a Game Piece
image

### Intake Configuration
image

### Ejector Configuration
image

### Temporary Components
These components are going to be left in and will be addressed shortly after in the scoring zone branch and likely one more branch to address input and behavior issues.

- Pieces are picked up via a ray cast from the intake zone, 3 meters forward along the z axis.
- Ejection is bound to left shift and control and isn't compatible with the InputSystem
- Pickup zones are always rendered
- There is no highlighting of rigid nodes

> [!Warning]
> Merge after #1007

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1728)

### CODE REVIEW

File: fission/src/mirabuf/MirabufSceneObject.ts
Comment (azaleacolburn) [inline fission/src/mirabuf/MirabufSceneObject.ts]: Why is this a function, but disabling physics for all rigid bodies in the same way isn't? It's not an issue, I'm just curious.

File: fission/src/systems/physics/PhysicsSystem.ts
Comment (azaleacolburn) [inline fission/src/systems/physics/PhysicsSystem.ts]: 5 lines of comments, 4 lines of code. This is what programming's all about. To be honest, these functions are completely self-documenting without any comments.

File: fission/src/systems/physics/PhysicsSystem.ts
Comment (azaleacolburn) [inline fission/src/systems/physics/PhysicsSystem.ts]: Utilize the `?` operator to reduce redundancy

```suggestion
            const vertArr = body.triangleMesh?.mesh?.verts
            const indexArr = body.triangleMesh?.mesh?.indices
            if (vertArr && indexArr) {
                for (let i = 0; i < vertArr.length; i += 3) {
                    const vert = MirabufFloatArr_JoltFloat3(vertArr, i)
                    settings.mTriangleVertices.push_back(vert)
                    this.UpdateMinMaxBounds(new JOLT.Vec3(vert), min, max)
                    JOLT.destroy(vert)
                }
                for (let i = 0; i < indexArr.length; i += 3) {
                    settings.mIndexedTriangles.push_back(
                        new JOLT.IndexedTriangle(indexArr.at(i)!, indexArr.at(i + 1)!, indexArr.at(i + 2)!, 0)
                    )
                }
            }
```

File: fission/src/systems/scene/SceneRenderer.ts
Comment (azaleacolburn) [inline fission/src/systems/scene/SceneRenderer.ts]: Extracting this operation to a variable might be nice

**Review by azaleacolburn — DISMISSED** [review]

File: fission/src/mirabuf/MirabufSceneObject.ts
Comment (Dhruv-0-Arora) [reply fission/src/mirabuf/MirabufSceneObject.ts]: I added that to solve a bug so that is a good point and ill add that. The problem is that all of that stuff is from PR #1009 so I might need to wait for that to be merged

File: fission/src/systems/physics/PhysicsSystem.ts
Comment (Dhruv-0-Arora) [reply fission/src/systems/physics/PhysicsSystem.ts]: I don't even know why I did that but I do know its gone from #1009 😇

File: fission/src/systems/scene/SceneRenderer.ts
Comment (Dhruv-0-Arora) [reply fission/src/systems/scene/SceneRenderer.ts]: That is true it would be more optimized

**Comment by Dhruv-0-Arora** [comment]
Notes for Reviewers
1. checkselectednode, listrobots, etc.. are duplicated functions. Where should I put them
2. Can I optimized this in checkSelectedNode()
``` ts
selectedRobot?.mirabufInstance?.parser.rigidNodes.forEach(rn => {
            if (World.PhysicsSystem.GetBody(selectedRobot.mechanism.GetBodyByNodeId(rn.id)!).GetID() === body.GetID()) {
                bodyAttachmentRef.current = body
                returnValue = true
            }
        })
```

File: fission/src/mirabuf/EjectableSceneObject.ts
Comment (azaleacolburn) [inline fission/src/mirabuf/EjectableSceneObject.ts:36]: With if statements like this one and the analogous one in Update, reversing the condition seems to be common practice to void code nesting. This isn't an issue here, but in other places (like Eject) you reverse the condition, so you might want to do the same here. Btw the reverse of `x && y` is `!x || !y`

File: fission/src/mirabuf/IntakeSensorSceneObject.ts
Comment (azaleacolburn) [inline fission/src/mirabuf/IntakeSensorSceneObject.ts]: Not sure if this is possible, but you might be able to switch the position of these two if statements, then reverse each condition in the deeply nested one. If you can't switch them then nvm.

File: fission/src/mirabuf/MirabufParser.ts
Comment (azaleacolburn) [inline fission/src/mirabuf/MirabufParser.ts]: You could use the [map higher order function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) instead to [convert an array into a map](https://stackoverflow.com/questions/75290595/convert-array-to-map-typescript) simplify these lines.

File: fission/src/systems/physics/PhysicsSystem.ts
Comment (azaleacolburn) [inline fission/src/systems/physics/PhysicsSystem.ts]: If you have to use unknown as a middleman, will this always be a safe cast given the class constraints on T?

**Review by azaleacolburn — COMMENTED** [review]
What's the plan for resolving this PR with preferences? Will there be another PR with an overlapping scope, or have they been completely bundled together?

**Comment by HunterBarclay** [comment]
@Ewie21 Preferences will be merged in first, then this branch, then finally Ayla's branch will have all the glue for everything.

File: fission/src/mirabuf/EjectableSceneObject.ts
Comment (HunterBarclay) [reply fission/src/mirabuf/EjectableSceneObject.ts:36]: Generally speaking, I see doing `if (condition) { early out }` to be indicating that the function should have never been ran and it is a safe guard against poor user usage. Doing `if (condition) { then do thing }` indicates more that the function should've been ran, regardless of the state of condition. This also leaves it open for other things outside of the if statement to be ran potentially.

**Comment by azaleacolburn** [comment]
> Notes for Reviewers
> 
> 1. checkselectednode, listrobots, etc.. are duplicated functions. Where should I put them
> 2. Can I optimized this in checkSelectedNode()
> 
> ```ts
> selectedRobot?.mirabufInstance?.parser.rigidNodes.forEach(rn => {
>             if (World.PhysicsSystem.GetBody(selectedRobot.mechanism.GetBodyByNodeId(rn.id)!).GetID() === body.GetID()) {
>                 bodyAttachmentRef.current = body
>                 returnValue = true
>             }
>         })
> ```

You could make it into a regular for loop and break, since all you'll do is set the same variables again. Also, could you just compare the ids directly, or are ids not consistent between Physics System and mechanism? I'm sure there are other things as well, but I haven't throughly read through the code yet.

File: fission/src/mirabuf/EjectableSceneObject.ts
Comment (HunterBarclay) [reply fission/src/mirabuf/EjectableSceneObject.ts:36]: And at the end of the day, it's up to the compiler and the operating system to determine how it performs the jump instruction and predictive statements, not which way you write the if statement.

**Comment by HunterBarclay** [comment]
> You could make it into a regular for loop and break, since all you'll do is set the same variables again. Also, could you just compare the ids directly, or are ids not consistent between Physics System and mechanism? I'm sure there are other things as well, but I haven't throughly read through the code yet.

@Ewie21 This no longer exists.

File: fission/src/mirabuf/MirabufParser.ts
Comment (HunterBarclay) [reply fission/src/mirabuf/MirabufParser.ts]: I assumed that wasn't a thing for TypeScript, thanks.

File: fission/src/systems/physics/PhysicsSystem.ts
Comment (HunterBarclay) [reply fission/src/systems/physics/PhysicsSystem.ts]: I'm fairly certain that this cast doesn't actually properly type check and I don't know how to fix it

File: fission/src/mirabuf/EjectableSceneObject.ts
Comment (azaleacolburn) [reply fission/src/mirabuf/EjectableSceneObject.ts:36]: Oh ya, I didn't mean it was an optimization at all, I just meant we should be consistent with which choices we make. I like what you just said about what they indicate, that works for me 👍

File: fission/src/systems/physics/PhysicsSystem.ts
Comment (HunterBarclay) [reply fission/src/systems/physics/PhysicsSystem.ts]: I originally had a separate BodyAssociated object for gamepieces, but literally not matter what I did, it would somehow successfully force cast a RigidNodeAssociate to GamePieceAssociate. I had required object types that weren't allowed to be null just fully forced to be null and I have no idea why. You can do normal `as` but then it throws an error on fail instead of just returning undefined.

Comment (azaleacolburn) [reply fission/src/systems/physics/PhysicsSystem.ts]: Ok, maybe make a note about what kind of types T work, if you have an idea that might be helpful to anyone else using this function.

File: fission/src/mirabuf/IntakeSensorSceneObject.ts
Comment (HunterBarclay) [reply fission/src/mirabuf/IntakeSensorSceneObject.ts]: I can switch the statements, but I'm not reversing the conditions as explained for the code in Ejectable

Comment (azaleacolburn) [reply fission/src/mirabuf/IntakeSensorSceneObject.ts]: Ya ignore this then 👍

**Comment by HunterBarclay** [comment]
Note: Rigid node highlighting can be implemented once #1023 is merged in.

File: fission/src/systems/physics/PhysicsSystem.ts
Comment (HunterBarclay) [reply fission/src/systems/physics/PhysicsSystem.ts]: Honestly the issue is with horrid generics support for TypeScript. I think I'm going to dispense with it and leave it to the user.

Comment (azaleacolburn) [reply fission/src/systems/physics/PhysicsSystem.ts]: Ok ya that sounds good. Generic traits in TS when?

**Comment by LucaHaverty** [comment]
I'm not sure when this was introduced, but opening the poker panel or settings modal throws this error. It might be anything that uses a slider.
```
TypeError: Cannot read properties of undefined (reading 'toLocaleString')
    at Slider (Slider.tsx:38:44)
    at renderWithHooks (react-dom.development.js:15486:18)
    at mountIndeterminateComponent (react-dom.development.js:20103:13)
    at beginWork (react-dom.development.js:21626:16)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:4164:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:4213:16)
    at invokeGuardedCallback (react-dom.development.js:4277:31)
    at beginWork$1 (react-dom.development.js:27490:7)
    at performUnitOfWork (react-dom.development.js:26596:12)
    at workLoopSync (react-dom.development.js:26505:5)
```

**Comment by HunterBarclay** [comment]
@LucaHaverty  Oops, I'll take care of it

File: fission/src/ui/components/Slider.tsx
Comment (LucaHaverty) [inline fission/src/ui/components/Slider.tsx:15]: Perhaps this is the cause of the slider issues? Maybe due to a merge?

**Review by LucaHaverty — COMMENTED** [review]
Besides the slider things, I could not find any other issues, great work!

File: fission/src/ui/components/Slider.tsx
Comment (Dhruv-0-Arora) [reply fission/src/ui/components/Slider.tsx:15]: @LucaHaverty could you look at that again and see if you can find errors with the fix
I didn't touch the ZoneConfigPanel because I don't want merge conflicts from your PR

**Review by LucaHaverty — APPROVED** [review]
Everything worked well, fantastic work!

File: fission/src/mirabuf/IntakeSensorSceneObject.ts
Comment (Dhruv-0-Arora) [inline fission/src/mirabuf/IntakeSensorSceneObject.ts]: could you do gpAssoc?.isGamePiece

File: fission/src/mirabuf/MirabufSceneObject.ts
Comment (Dhruv-0-Arora) [inline fission/src/mirabuf/MirabufSceneObject.ts]:
```suggestion
import { BodyAssociate, LayerReserve } from "@/systems/physics/PhysicsSystem"
```

**Review by BrandonPacewic — COMMENTED** [review]
@Dhruv-0-Arora Are you not the author of this PR? Or did @HunterBarclay take this one over?

**Comment by Dhruv-0-Arora** [comment]
> @Dhruv-0-Arora Are you not the author of this PR? Or did @HunterBarclay take this one over?

Hunter is the owner for this one

**Review by a-crowell — APPROVED** [review]
Works great! Just removed an unused import.

File: fission/src/mirabuf/MirabufSceneObject.ts
Comment (a-crowell) [reply fission/src/mirabuf/MirabufSceneObject.ts]: Fixed

**Review by BrandonPacewic — CHANGES_REQUESTED** [review]
Want to make sure @Dhruv-0-Arora 's comments are looked at before merge.

> could you do gpAssoc?.isGamePiece
> 
> _Originally posted by @Dhruv-0-Arora in https://github.com/Autodesk/synthesis/pull/1017#discussion_r1678356692_
>
> @LucaHaverty could you look at that again and see if you can find errors with the fix
I didn't touch the ZoneConfigPanel because I don't want merge conflicts from your PR
> 
> _Originally posted by @Dhruv-0-Arora in https://github.com/Autodesk/synthesis/pull/1017#discussion_r1678271072_
>
> That is true it would be more optimized
>
> _Originally posted by @Dhruv-0-Arora in https://github.com/Autodesk/synthesis/pull/1017#discussion_r1671662056_

File: fission/src/mirabuf/IntakeSensorSceneObject.ts
Comment (HunterBarclay) [reply fission/src/mirabuf/IntakeSensorSceneObject.ts]: Yes I believe so

**Review by BrandonPacewic — APPROVED** [review]

**Pull request merged**
